File: BaekjoonAlgorithm/BA_6593/BJAlgo_6593.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.Comparator;
import java.util.PriorityQueue;
import java.util.StringTokenizer;

public class BJAlgo_6593 {
    private static int[] dz = {-1, 1, 0, 0, 0, 0};
    private static int[] dx = {0, 0, -1, 1, 0, 0};
    private static int[] dy = {0, 0, 0, 0, -1, 1};

    private static int l;
    private static int r;
    private static int c;

    private static String[][] building;

    private static int sz;
    private static int sx;
    private static int sy;
    private static int ez;
    private static int ex;
    private static int ey;


    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;

        while (true) {
            st = new StringTokenizer(br.readLine());
            l = Integer.parseInt(st.nextToken());
            r = Integer.parseInt(st.nextToken());
            c = Integer.parseInt(st.nextToken());

            if (l == 0 && r == 0 && c == 0) {
                break;
            }

            building = new String[l][r];
            for (int i = 0; i < l; i++) {
                for (int j = 0; j < r; j++) {
                    String tmp = br.readLine();
                    for (int k = 0; k < tmp.length(); k++) {
                        if (tmp.charAt(k) == 'S') {
                            sz = i;
                            sx = j;
                            sy = k;
                        } else if (tmp.charAt(k) == 'E') {
                            ez = i;
                            ex = j;
                            ey = k;
                        }
                    }
                    building[i][j] = tmp;
                }
                br.readLine();
            }

            int result = bfs();
            if (result != 0) {
                System.out.println("Escaped in " + result + " minute(s).");
            } else {
                System.out.println("Trapped!");
            }
        }

    }

    private static int bfs() {
        PriorityQueue<int[]> queue = new PriorityQueue<>(new Comparator<int[]>() {
            @Override
            public int compare(int[] t1, int[] t2) {
                return t1[0] - t2[0];
            }
        });
        queue.offer(new int[]{0, sz, sx, sy});

        boolean[][][] visited = new boolean[l][r][c];
        visited[sz][sx][sy] = true;

        while (!queue.isEmpty()) {
            int[] current = queue.poll();

            if (current[1] == ez && current[2] == ex && current[3] == ey) {
                return current[0];
            }

            for (int d = 0; d < 6; d++) {
                int nz = current[1] + dz[d];
                int nx = current[2] + dx[d];
                int ny = current[3] + dy[d];

                if (nz >= 0 && nz < l && nx >= 0 && nx < r && ny >= 0 && ny < c) {
                    if (building[nz][nx].charAt(ny) != '#' && !visited[nz][nx][ny]) {
                        queue.offer(new int[]{current[0] + 1, nz, nx, ny});
                        visited[nz][nx][ny] = true;
                    }
                }
            }
        }

        return 0;
    }
}

File: BaekjoonAlgorithm/202101/BA_6593/BJAlgo_6593.py
from sys import stdin
import heapq

dx = [-1, 1, 0, 0, 0, 0]
dy = [0, 0, -1, 1, 0, 0]
dz = [0, 0, 0, 0, -1, 1]


def bfs(z, x, y):
    queue = []
    heapq.heappush(queue, [0, z, x, y])

    visited = [[[False for _ in range(c)] for _ in range(r)] for _ in range(l)]
    visited[z][x][y] = True

    while queue:
        cnt, cz, cx, cy = heapq.heappop(queue)

        if cz == ez and cx == ex and cy == ey:
            return cnt

        for d in range(6):
            nz = cz + dz[d]
            nx = cx + dx[d]
            ny = cy + dy[d]

            if 0 <= nz < l and 0 <= nx < r and 0 <= ny < c:
                if building[nz][nx][ny] != '#' and not visited[nz][nx][ny]:
                    heapq.heappush(queue, [cnt + 1, nz, nx, ny])
                    visited[nz][nx][ny] = True

    return 0


while True:
    l, r, c = map(int, stdin.readline().split())

    if l == r == c == 0:
        break

    sx = 0
    sy = 0
    sz = 0

    ex = 0
    ey = 0
    ez = 0

    building = [[] * r for _ in range(l)]
    for i in range(l):
        for j in range(r):
            tmp = list(map(str, stdin.readline().strip()))
            if 'S' in tmp:
                sz = i
                sx = j
                sy = tmp.index('S')
            elif 'E' in tmp:
                ez = i
                ex = j
                ey = tmp.index('E')
            building[i].append(tmp)
        stdin.readline()

    result = bfs(sz, sx, sy)
    if result == 0:
        print("Trapped!")
    else:
        print("Escaped in", result, "minute(s).")
